### PR TITLE
JAMES-3441 Avoid starting mail spooler when 'threads' = 0 

### PR DESCRIPTION
**Pull request by chibenwa**
This allows configuring which James instances, within a distributed topology, should
handle mail processing.

### CODE REVIEW

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java [hunk 1]
```diff
@@ -227,12 +229,17 @@ public void configure(HierarchicalConfiguration<ImmutableNode> config) {
...
+        Preconditions.checkArgument(concurrencyLevel >= 0, "'threads' needs to be greater than or equal to zero");
+        if (concurrencyLevel > 0) {
+            LOGGER.info("init...");
+            LOGGER.info("Concurrency level is {}", concurrencyLevel);
+            MailQueue queue = queueFactory.createQueue(MailQueueFactory.SPOOL, MailQueueFactory.prefetchCount(concurrencyLevel));
+            runner = Optional.of(new Runner(metricFactory,
+                mailProcessor, errorRepository(), errorRepositoryURL, queue, concurrencyLevel));
+            LOGGER.info("Spooler started");
+        } else {
+            LOGGER.info("Spooler had been dis-activated. To enable is set 'threads' count to a value greater than zero");
```
Comment (Arsnael) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java:282]: s/dis-activated/disactivated

Comment (Arsnael) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java:282]: s/is/it

**Review by Arsnael — APPROVED** [review]

File: server/container/guice/memory-guice/src/test/java/org/apache/james/DisabledSpoolerTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            .addUser(ALICE.asString(), ALICE_PASSWORD)
+            .addUser(BOB.asString(), BOB_PASSWORD);
+    }
+
+    @Test
+    void emailsShouldNotBeDequeuedWhenZeroSpoolerThreads(GuiceJamesServer server) throws Exception {
+        smtpMessageSender.connect(LOCALHOST_IP, server.getProbe(SmtpGuiceProbe.class).getSmtpPort())
+            .sendMessageWithHeaders(ALICE.asString(), BOB.asString(), MESSAGE);
+
+        Thread.sleep(5000);
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/container/guice/memory-guice/src/test/java/org/apache/james/DisabledSpoolerTest.java:77]: No, we want to check here that the interaction do not happen. It needs to check after some time.

**Comment by rouazana** [comment]
> > is it well documented that 0 means really 0 threads and is not a magic value for example telling that it could be infinite?
> 
> To be handled in #4048

Then the ticket is not precise enough. My comment was about checking existing documentation. For my part when you modify the behavior of the server regarding a configuration option, you should modify the matching documentation at the same time.

**Review by Arsnael — APPROVED** [review]

**Review by LanKhuat — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**